File: shortest_word_distance3.cpp
// Time Complexity : O(n)
// Space Complexity : O(1)
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
//1. crete two pointers to point to first and second word.
//2. If two words are equal then i1 will be same as the index.
//3. change i1 value to i2
//4. update minimum distance at each find

class Solution {
public:
    int shortestWordDistance(vector<string>& words, string word1, string word2) {
        int size = words.size();
        int i1=-1, i2=-1;
        int min_dist = INT_MAX;
        for(int i=0; i<size; i++){
            if(words[i].compare(word1)==0){
                i1=i;
            }
            if(words[i].compare(word2)==0){
                if(i1==i)
                    i1=i2;
                i2=i;
            }
            if(i1 !=-1 && i2 != -1){
                min_dist = min(min_dist, abs(i1-i2));
            }
        }
        return min_dist;
    }
};

File: ShortestWordDistance3.java
// Time Complexity: O(N)
// Space Complexity: O(1)
class Solution {
    public int shortestWordDistance(String[] wordsDict, String word1, String word2) {
        if (wordsDict == null || wordsDict.length == 0) return 0;

        int i1=-1, i2=-1;
        int min = Integer.MAX_VALUE;

        for(int i=0; i < wordsDict.length; i++) {
            String word = wordsDict[i];
            if (word.equals(word1)) {
                i1=i;
            }

            if(word.equals(word2)) {
                if (i1 == i) {
                    i1=i2;
                }
                i2=i;
            }
            if (i1 != -1 && i2 != -1) {
                min = Math.min(min, Math.abs(i1 -i2));
            }

        }

        // if (!word1.equals(word2)) {
        //     for(int i=0; i< wordsDict.length; i++) {
        //         String word = wordsDict[i];
        //         if (word.equals(word1)) {
        //             i1=i;
        //         }
        //         if (word.equals(word2)) {
        //             i2=i;
        //         }
        //         if (i1 != -1 && i2 != -1) {
        //             min = Math.min(min, Math.abs(i1 -i2));
        //         }
        //     }
        // } else {
        //     for(int i=0; i< wordsDict.length; i++) {
        //         String word = wordsDict[i];
        //         if (word.equals(word1)) {
        //             if (i1 < i2) {
        //                 i1 = i;
        //             } else {
        //                 i2 = i;
        //             }
        //         }
        //         if (i1 != -1 && i2 != -1) {
        //             min = Math.min(min, Math.abs(i1 -i2));
        //         }
        //     }
        // }

        return min;
    }
}